File: archive/p/python/bubble-sort.py
#!/usr/bin/env python
import sys
import argparse
from functools import reduce


def bubble_sort(xs):
    def pass_list(xs):
        if len(xs) <= 1:
            return xs
        x0 = xs[0]
        x1 = xs[1]
        if x1 < x0:
            del xs[1]
            return [x1] + pass_list(xs)
        return [x0] + pass_list(xs[1:])
    return reduce(lambda acc, _ : pass_list(acc), xs, xs[:])


def input_list(list_str):
    return [int(x.strip(" "), 10) for x in list_str.split(',')]


if __name__ == "__main__":
    usage = 'Usage: please provide a list of at least two integers to sort in the format “1, 2, 3, 4, 5”'
    parser = argparse.ArgumentParser(usage=usage)
    parser.add_argument('list', type=input_list)
    args = parser.parse_args()
    if len(args.list) <= 1:
        print(usage)
        sys.exit(1)

    print(bubble_sort(args.list))

File: archive/p/pascal/bubble_sort.pas
program BubbleSort;

{$mode objfpc}{$H+}

uses
   Classes, Generics.Collections, SysUtils;

type
   TIntegerList = specialize TList<Integer>;

procedure ShowUsage;
begin
   Writeln('Usage: please provide a list of at least two integers to sort in the format "1, 2, 3, 4, 5"');
   Halt(1);
end;

function ParseIntegerList(const S: string): TIntegerList;
var
   Tokens: TStringArray;
   Token: string;
   Value: Integer;
begin
   if S.Trim.IsEmpty then
      ShowUsage;

   Tokens := S.Split([',']);
   Result := TIntegerList.Create;

   for Token in Tokens do
   begin
      if not TryStrToInt(Trim(Token), Value) then
      begin
         Result.Free;
         ShowUsage;
      end;
      Result.Add(Value);
   end;

   if Result.Count < 2 then
   begin
      Result.Free;
      ShowUsage;
   end;
end;

function IsSorted(const A: TIntegerList): Boolean;
var
   i: Integer;
begin
   for i := 1 to A.Count - 1 do
      if A[i] < A[i - 1] then
         Exit(False);
   Result := True;
end;

procedure BubbleSort(List: TIntegerList);
var
   i, j: Integer;
   Swapped: Boolean;

   procedure SwapElements(Index1, Index2: Integer);
   var
      Temp: Integer;
   begin
      Temp := List[Index1];
      List[Index1] := List[Index2];
      List[Index2] := Temp;
   end;
begin
   for i := List.Count - 1 downto 1 do
   begin
      Swapped := False;
      for j := 0 to i - 1 do
         if List[j] > List[j + 1] then
         begin
            SwapElements(j, j + 1);
            Swapped := True;
         end;

      if not Swapped then
         Break;
   end;
end;

function FormatIntegerList(List: TIntegerList): string;
var
   i: Integer;
begin
   Result := '';
   for i := 0 to List.Count - 1 do
   begin
      if i > 0 then
         Result += ', ';
      Result += IntToStr(List[i]);
   end;
end;

var
   RawInput: string;
   Numbers: TIntegerList;
begin
   if ParamCount <> 1 then
      ShowUsage;

   RawInput := ParamStr(1);
   Numbers := ParseIntegerList(RawInput);
   try
      if not IsSorted(Numbers) then
         BubbleSort(Numbers);

      Writeln(FormatIntegerList(Numbers));
   finally
      Numbers.Free;
   end;
end.

